### PR TITLE
block_synchronizer: add a new final state and fix signature sufficiency determination for blocks with pre-1.5 protocol versions

### LINKED ISSUE #3824: Fix determination for signature sufficiency in the block synchronizer
When we try to forward sync a block that doesn’t have deploys, the builder can switch to execute the block even if it doesn’t have strict finality. The bug is exposed by the having_block_body_for_block_without_deploys_requires_only_signatures test in [this PR](https://github.com/casper-network/casper-node/pull/3816). This is because [here](https://github.com/casper-network/casper-node/blob/e5d21032f5590648315787a54546eea6c43e8b5c/node/src/components/block_synchronizer/block_acquisition_action.rs#L324) we don’t first check if this is a forward block and since is_checkable is only set on the historical sync path, we fall through and will report finality based on LegacyRequiredFinality which is not the intended behaviour. Furthermore currently we allow weak finality to be sufficient for checkable blocks when doing a historical sync.

What we actually want is:
- Forward blocks -> always require strict finality
- Historical legacy blocks -> determination based on LegacyRequiredFinality
- Historical checkable blocks -> always require strict finality

**Update:** The current code relaxes the signature finality sufficiency criteria only for the case where we want Strict finality (using LegacyRequiredFinality). But the first milestone when synchronizing the block is to reach Weak finality. In the case where LegacyRequiredFinality::Any is set for a network where weak finality signatures cannot be fetched, the synchronizer will be stuck trying to get Weak finality since the requirement is not relaxed with the current code. [This PR](https://github.com/casper-network/casper-node/pull/3822) does not fully address the issues described here.

### PR DESCRIPTION
**Pull request by alsrdn**
Fixes: https://github.com/casper-network/casper-node/issues/3824

`HaveStrictFinality` was considered the terminal state when synchronizing a block even though this was not entirely true. Because on the forward sync flow the synchronizer needed to wait until the block was executed to establish that the sync was finished, there was some unnecessary complexity added to keep `HaveStrictFinality` as a terminal state which involved more complex checks in other intermediary states like `HaveBlock`, `HaveAllDeploys` etc.
The same can be said about the historical sync flow where the synchronizer needed to wait for the block to be marked complete.
This PR introduces a new terminal block acquisition state called `Complete` which fully denotes that the block synchronizer is finished with all tasks involving the block, on both forward and historical sync.
This simplifies some state transitions since enqueuing the block for execution or requesting the block to be marked complete is only done from the `HaveStrictFinality` state.
For the forward sync case, there is another intermediary state called `HaveFinalizedBlock`; the block builder transitions to this state after the block has been successfully converted into a block that is ready for execution.

Furthermore, we now allow sync to progress for historical blocks that were created with a protocol version that is older than the one where the requirement for nodes to hold signatures for strict finality was introduced.
In mainnet this is protocol version 1.5.
For other networks a configurable chainspec parameter was added to control the version when this requirement was introduced.
Because pre-1.5 there was no requirement for nodes to retain finality signatures, there may exist networks were all finality signatures for a block are not available in order to reach weak or strict finality.
Because we have 2 checkpoints for determining if we can progress with the sync (first achieving weak finality, and then strict finality) we employ the `LegacyRequiredFinality` the requirement for the block that is within the criteria described above. Previous code did not take into account the check for weak finality.

### CODE REVIEW

**Review by Fraser999 — APPROVED** [review]

File: node/src/components/block_synchronizer/block_acquisition.rs [hunk 1]
```diff
@@ -509,6 +574,47 @@ impl BlockAcquisitionState {
     }
 
     /// Register a finality signature as pending for this block.
+    pub(super) fn switch_to_have_strict_finality(
```
Comment (rafal-ch) [inline node/src/components/block_synchronizer/block_acquisition.rs:580]: Looks like this comment and the one for the function below need to switch places.

File: node/src/components/block_synchronizer/block_acquisition.rs [hunk 2]
```diff
@@ -438,6 +495,10 @@ impl BlockAcquisitionState {
                         block_hash,
                         header.height()
                     );
+                    if is_historical && header.protocol_version() < strict_finality_protocol_version
+                    {
```
Comment (rafal-ch) [inline node/src/components/block_synchronizer/block_acquisition.rs]: This `{` in a new line reminds me of the code-formatting wars :-)

File: node/src/components/block_synchronizer/signature_acquisition.rs [hunk 3]
```diff
@@ -93,21 +100,45 @@ impl SignatureAcquisition {
...
         self.signature_weight
     }
+
+    // Determines signature weight sufficiency based on the type of sync (forward or historical) and
+    // the protocol version that the block was created with (pre-1.5 or post-1.5)
+    // `requires_strict_finality` determines what the caller requires with regards to signature
+    // sufficiency:
+    //      * false means that the caller considers `Weak` finality as sufficient
+    //      * true means that the caller considers `Strict` finality as sufficient
+    pub(super) fn has_sufficient_finality(
```
Comment (rafal-ch) [inline node/src/components/block_synchronizer/signature_acquisition.rs:121]: Can you cover this function with UT?

File: node/src/components/block_synchronizer/block_acquisition.rs (hunk 2 above)
Comment (alsrdn) [reply node/src/components/block_synchronizer/block_acquisition.rs]: fixed as part of: https://github.com/casper-network/casper-node/pull/3832/commits/fc3e6afc1098cca19300d76e87f4595853042470

File: node/src/components/block_synchronizer/block_acquisition.rs (hunk 1 above)
Comment (alsrdn) [reply node/src/components/block_synchronizer/block_acquisition.rs:580]: fixed as part of: https://github.com/casper-network/casper-node/pull/3832/commits/fc3e6afc1098cca19300d76e87f4595853042470

**Review by georgepisaltu — APPROVED** [review]

File: node/src/components/block_synchronizer.rs [hunk 4]
```diff
@@ -1252,15 +1252,29 @@ impl BlockSynchronizer {
...
+                last_progress_time,
+            )
+        } else {
+            BlockSynchronizerProgress::Syncing(
+                builder.block_hash(),
+                builder.block_height(),
+                builder.last_progress_time().max(
+                    self.global_sync
+                        .last_progress()
+                        .unwrap_or_else(Timestamp::zero),
```
Comment (georgepisaltu) [inline node/src/components/block_synchronizer.rs]: Could have used `last_progress_time` here too.

Comment (rafal-ch) [reply node/src/components/block_synchronizer.rs]: Addressed in 961750e61cca7014966adcf3785324191038f2f7

**Review by georgepisaltu — APPROVED** [review]

File: node/src/components/block_synchronizer/signature_acquisition.rs (hunk 3 above)
Comment (rafal-ch) [reply node/src/components/block_synchronizer/signature_acquisition.rs:121]: This is covered in a separate PR: https://github.com/casper-network/casper-node/pull/3842

**Pull request merged**